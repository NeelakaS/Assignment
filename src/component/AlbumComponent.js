import React, {useEffect} from 'react';
import {
  View,
  Text,
  FlatList,
  TouchableOpacity,
  ActivityIndicator,
} from 'react-native';
import {connect} from 'react-redux';
import {fetchAlbumData} from '../action/actions';
import styles from './AlbumComponent.style';

type AlbumData = {
  userId: number,
  id: number,
  title: string,
};

type Props = {
  fetchAlbumInfo: (userId: string) => void,
  albumData: AlbumData,
  navigation: Objct,
  loading: boolean,
};

const AlbumComponent = (props: Props) => {
  const {fetchAlbumInfo, albumData, navigation, loading} = props;
  const userId = navigation.getParam('userId');

  // hooks
  useEffect(() => {
    fetchAlbumInfo(userId);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [userId]);

  const extractKey = (item: AlbumData, index: number) => `${item?.id}-${index}`;

  const renderItem = ({item}: AlbumData) => {
    return (
      <TouchableOpacity
        style={styles.listContainer}
        onPress={() => {
          navigation.navigate('Photo', {
            albumId: item.id,
          });
        }}>
        <Text style={styles.albumTitle} numberOfLines={2}>
          {item.title}
        </Text>
      </TouchableOpacity>
    );
  };

  // loading untill API receives data
  if (loading) {
    return (
      <View style={styles.actionIndicator}>
        <ActivityIndicator />
      </View>
    );
  }

  return (
    <FlatList
      ItemSeparatorComponent={() => <View style={styles.listItemSeparator} />}
      data={albumData}
      renderItem={renderItem}
      keyExtractor={extractKey}
    />
  );
};

const mapDispatchToProps = dispatch => ({
  fetchAlbumInfo: userId => dispatch(fetchAlbumData(userId)),
});
const mapStateToProps = state => ({
  albumData: state.albumData.data,
  loading: state.albumData.isFetching,
});

// eslint-disable-next-line prettier/prettier
export default connect(mapStateToProps, mapDispatchToProps)(AlbumComponent);
