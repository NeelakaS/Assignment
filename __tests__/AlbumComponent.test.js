import 'react-native';
import React from 'react';

import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import saga from 'redux-saga';

import AlbumComponent from '../src/component/AlbumComponent';

const ALBUM_DATA = [
  {
    userId: 9,
    id: 82,
    title: 'recusandae consequatur vel amet unde',
  },
  {
    userId: 9,
    id: 83,
    title: 'aperiam odio fugiat',
  },
  {
    userId: 9,
    id: 84,
    title: 'est et at eos expedita',
  },
];

const DEFAULT_PROPS = {
  fetchUserData: jest.fn(),
  albumData: [],
  navigation: {
    getParam: jest.fn(),
  },
  loading: false,
};

const middlewares = [saga];
const mockStore = configureMockStore(middlewares);
let wrapper;

describe('User Compoent', () => {
  it('render UserComponent correctly', () => {
    const initialState = {
      albumData: {
        data: ALBUM_DATA,
        isDataFetched: false,
        isFetching: false,
        isError: false,
      },
    };
    const store = mockStore(initialState);

    wrapper = renderer
      .create(
        <Provider store={store}>
          <AlbumComponent {...DEFAULT_PROPS} />
        </Provider>,
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
