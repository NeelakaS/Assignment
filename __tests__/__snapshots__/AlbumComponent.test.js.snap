// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album Component render AlbumComponent correctly 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={
    Array [
      Object {
        "id": 82,
        "title": "recusandae consequatur vel amet unde",
        "userId": 9,
      },
      Object {
        "id": 83,
        "title": "aperiam odio fugiat",
        "userId": 9,
      },
      Object {
        "id": 84,
        "title": "est et at eos expedita",
        "userId": 9,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#ffffff",
            "flex": 1,
            "height": 60,
            "justifyContent": "center",
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          numberOfLines={2}
          style={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
              "lineHeight": 25,
            }
          }
        >
          recusandae consequatur vel amet unde
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#404041",
            "minHeight": 0.5,
            "opacity": 0.3,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#ffffff",
            "flex": 1,
            "height": 60,
            "justifyContent": "center",
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          numberOfLines={2}
          style={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
              "lineHeight": 25,
            }
          }
        >
          aperiam odio fugiat
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#404041",
            "minHeight": 0.5,
            "opacity": 0.3,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#ffffff",
            "flex": 1,
            "height": 60,
            "justifyContent": "center",
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          numberOfLines={2}
          style={
            Object {
              "fontSize": 15,
              "fontWeight": "bold",
              "lineHeight": 25,
            }
          }
        >
          est et at eos expedita
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
